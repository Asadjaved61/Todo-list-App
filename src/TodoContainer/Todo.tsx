import React, { useState } from 'react'
import TodoProps from '../Interfaces/TodoProps.interface';

const Todo: React.FC<TodoProps> = ({ todo, index, updateTodoStatus, removeTodo }) => {
    const [ edit, toggleEdit ] = useState<boolean>(false);

    const updateTodoText = (e: React.ChangeEvent<HTMLInputElement>) => {
        // update todo text by gettign input value
        todo.text = e.target.value
    }

    return (
        <li className="todo center-element" style={{ background: todo.isCompleted ? '#ccc' : 'white' }}>
            { edit ? 
                <input type="text" 
                    defaultValue={todo.text} 
                    className="input input-todoText"
                    name="todoText" 
                    onChange={updateTodoText} autoFocus/> : 
                <span 
                    className="todo-text" 
                    style={{ textDecoration: todo.isCompleted ? 'line-through': '' }}>{todo.text}
                </span> }
            <button className="btn close-btn" onClick={() => removeTodo(index)}>X</button>
            <button className="btn mark-complete-btn" onClick={ () => updateTodoStatus(index)}>
                { todo.isCompleted ? "Undone" : "Done" }
            </button>
            <button
                    className="btn edit-btn"
                    style={{ display: todo.isCompleted ? "none" : "inline" }} 
                    onClick={() => toggleEdit(!edit)}>
                { edit ? "Save" : "Edit" }
            </button>
        </li>
    )
}

export default Todo;