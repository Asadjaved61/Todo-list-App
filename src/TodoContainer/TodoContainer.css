.center-element {
    position: relative;
    left: 50%;
    transform: translate(-50%, 0%);
}

.todo-list-container {
    width: 50%;
    min-height: 200px;
    max-height: 700px;
    overflow: auto;
    background: rgb(116, 17, 74);
    margin-top: 5%;
    padding-bottom: 20px
}

.todo-list {
    margin: 0;
    padding: 0;
    list-style-type: none;
}

.todo {
    width: 95%;
    background: white;
    margin-top: 10px;
    font-size: 18px;
    padding: 10px;
    border-radius: 5px;
    color: #333;
    box-shadow: 1px 1px 1px rgba(0, 0, 0, 0.15);
}

.todo-text {
    display: inline-block;
    word-break: break-all;
    width: 60%
}

.btn {
    float: right;
    padding: 3px 7px;
    outline: none;
    border: none;
    background: none;
    border-radius: 3px;
    font-weight: 500;
    color: white;
    margin-left: 3px
}

.mark-complete-btn {
    background: rgb(24, 233, 129)
}

.mark-complete-btn:hover {
    background: rgb(40, 189, 115)
}

.close-btn {
    background: rgb(233, 120, 100);
}

.close-btn:hover {
    background: rgb(228, 90, 66);
}

.edit-btn {
    background: dodgerblue;
}

.edit-btn:hover {
    background: rgb(2, 107, 212);
}

.add-todo-form {
    width: 70%;
    position: relative;
    margin-top: 10px;
    margin-bottom: 30px
}

.input {
    padding: 8px;
    font-size: 16px;
    border: none;
    outline: none;
    background: none;
    border-bottom: 1px solid white;
}

.todo .input {
    border-bottom: 1px solid #999
}

.input-todoText {
    width: 60%
}

.add-todo-form .input {
    width: 80%;
    color: white
}

.add-todo-form .btn-submit {
    font-size: 16px;
    width:20%;
    padding: 8px 10px;
    background: yellowgreen;
    margin: 0;
}

@media only screen and (max-width: 800px) {
    .todo-list-container {
        width: 80%
    }

    .todo-text {
        width: 100%
    }

    .input-todoText {
        width: 100%
    }

    .btn {
        display: inline-block;
        float: none;
    }

    .add-todo-form {
        width: 100%
    }

    .add-todo-form .btn-submit {
        width: 20%;
        font-size: 12px
    }
}