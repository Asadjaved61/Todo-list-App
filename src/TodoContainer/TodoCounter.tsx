import React from 'react'

const TodoCounter = ({ todoCount }: any) => {
    return (
        <span className="todo-counter">
            { todoCount }
        </span>
    )
}

export default TodoCounter;
