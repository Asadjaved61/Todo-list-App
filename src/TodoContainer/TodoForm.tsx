import React, { useState } from 'react';
import TodoFormProps from '../Interfaces/TodoForm.interface';

const TodoForm: React.FC<TodoFormProps> = ({ addTodo }) => {
    const [value, setValue] = useState<string>("");
  
    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      if (!value) return;

      addTodo(value);
    };
  
    return (
      <form className="add-todo-form center-element" onSubmit={handleSubmit}>
        <input
          type="text"
          className="input"
          placeholder="Add todo..."
          value={value}
          onChange={e => setValue(e.target.value)}
          required
        />
        <button type="submit" className="btn btn-submit">Add</button>
      </form>
    );
  }

  export default TodoForm;