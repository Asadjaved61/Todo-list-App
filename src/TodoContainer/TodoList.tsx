import React from 'react';
import Todo from './Todo';

import './TodoContainer.css';
import TodoListProps from '../Interfaces/TodoListProps.interface';
import TodoValue from '../Interfaces/TodoValue.interface';

const TodoList: React.FC<TodoListProps> = ({ todos, updateTodoStatus, removeTodo }) => {
    return (
        <ul className = "todo-list">
            {todos.map((todo: TodoValue, index: number) => {
                return (
                    <Todo 
                        key= {index}
                        todo= {todo}
                        index= {index}
                        updateTodoStatus= {updateTodoStatus}
                        removeTodo= {removeTodo}
                    />
                )
            })}
        </ul>
    )
}

export default TodoList;
