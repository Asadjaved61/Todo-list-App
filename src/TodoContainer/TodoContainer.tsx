import React, {useState} from 'react'

import TodoForm from './TodoForm';
import TodoList from './TodoList';
import TodoValue from '../Interfaces/TodoValue.interface';
import TodoCounter from './TodoCounter';

const TodoContainer: React.FC = () => {
    const [todos, setTodos] = useState<TodoValue[]>([
        { text: 'Go to gym', isCompleted: false }
    ]);
    
    const addTodo = (text: String) => {
        const newTodos: any = [...todos, { text, isCompleted: false }];

        setTodos(newTodos);
    }

    const todoCount = () => {
        let count = 0;

        todos.forEach(todo => {
            if (!todo.isCompleted) {
                count += 1
            }
        })

        return count;
    }

    const moveDoneTaskAtBottom = () => {
        const newTodos: TodoValue[] = [...todos];

        newTodos.forEach((todo: TodoValue, index: number) => {
            if (todo.isCompleted) {
                newTodos.push(newTodos.splice(index, 1)[0])
            } else {
                newTodos.unshift(newTodos.splice(index, 1)[0])
            }
        })

        setTodos(newTodos)
    }

    const updateTodoStatus = (index: any) => {
        const newTodos: TodoValue[] = [...todos];

        newTodos[index].isCompleted = !newTodos[index].isCompleted;
        setTodos(newTodos);

        moveDoneTaskAtBottom()
    }

    const removeTodo = (index: any) => {
        const newTodos: TodoValue[] = [...todos];
        newTodos.splice(index, 1);

        setTodos(newTodos);
    }
    return (
        <div className="todo-list-container center-element">
            <TodoCounter todoCount= {todoCount()}/>
            <TodoForm addTodo= {addTodo}/>
            <TodoList 
                todos= {todos}
                updateTodoStatus= {updateTodoStatus}
                removeTodo= {removeTodo}
            />
        </div>
    )
}

export default TodoContainer;
