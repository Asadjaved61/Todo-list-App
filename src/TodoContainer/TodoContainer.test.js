import React from 'react'
import {render, fireEvent, cleanup, getByText} from '@testing-library/react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TodoContainer from './TodoContainer';

describe('Todo Container', () => {
    Enzyme.configure({ adapter: new Adapter() })
    afterEach(cleanup);

    it('it should match snapshot', () => {
        const { asFragment } = render(<TodoContainer />);

        expect(asFragment()).toMatchSnapshot();
    })

    it('addTodo should add new todo', () => {
       const wrapper = mount(<TodoContainer />);
       wrapper.instance().addTodo('shopping');

       const expectedValue = wrapper.instance().state.todos[0];
       expect(expectedValue).toMatchObject({ text: 'shopping', isCompleted: false })
    })

    it('moveDoneTaskAtBottom should move completed task to bottom', async () => {
        const wrapper = mount(<TodoContainer />);

        wrapper.instance().addTodo('shopping');
        wrapper.instance().addTodo('buy food');
        wrapper.instance().updateTodoStatus(0);
        wrapper.instance().moveDoneTaskAtBottom();

        expect(wrapper.instance().state.todos[1].text).toEqual('shopping')
     })

     it('updateTodoStatus should update todo status', async () => {
        const wrapper = mount(<TodoContainer />);

        wrapper.instance().addTodo('shopping');
        wrapper.instance().updateTodoStatus(0);

        expect(wrapper.instance().state.todos[0].isCompleted).toBeTruthy();
     })

     it('removeTodo should remove todo', async () => {
        const wrapper = mount(<TodoContainer />);

        wrapper.instance().addTodo('shopping');
        wrapper.instance().removeTodo(0);

        expect(wrapper.instance().state.todos[0]).toBeUndefined()
     })

     it('todoCount should give number of todos', async () => {
        const wrapper = mount(<TodoContainer />);

        wrapper.instance().addTodo('shopping');
        wrapper.instance().addTodo('buy food');

        const expectedValue = wrapper.instance().todoCount();

        expect(expectedValue).toBe(2)
     })
});