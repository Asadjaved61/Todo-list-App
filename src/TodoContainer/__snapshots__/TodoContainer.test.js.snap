// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Container it should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="todo-list-container center-element"
  >
    <span
      class="todo-counter"
    >
      0
    </span>
    <form
      class="add-todo-form center-element"
    >
      <input
        class="input add-todo-input"
        name="addTodo"
        placeholder="Add todo..."
        required=""
        type="text"
        value=""
      />
      <button
        class="btn btn-submit"
        type="submit"
      >
        Add
      </button>
    </form>
    <ul
      class="todo-list"
    />
  </div>
</DocumentFragment>
`;
